File: georocket-server/src/test/java/io/georocket/output/AllSameStrategyTest.java
package io.georocket.output;

import java.util.Arrays;

import org.junit.Test;
import org.junit.runner.RunWith;

import io.georocket.BufferWriteStream;
import io.georocket.storage.ChunkMeta;
import io.georocket.util.XMLStartElement;
import io.georocket.util.io.DelegateChunkReadStream;
import io.vertx.core.buffer.Buffer;
import io.vertx.ext.unit.Async;
import io.vertx.ext.unit.TestContext;
import io.vertx.ext.unit.junit.VertxUnitRunner;

/**
 * Test {@link AllSameStrategy}
 * @author Michel Kraemer
 */
@RunWith(VertxUnitRunner.class)
public class AllSameStrategyTest {
  private static final String XMLHEADER = "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n";
  
  private Buffer chunk1 = Buffer.buffer(XMLHEADER + "<root><test chunk=\"1\"></test></root>");
  private Buffer chunk2 = Buffer.buffer(XMLHEADER + "<root><test chunk=\"2\"></test></root>");
  private ChunkMeta cm = new ChunkMeta(Arrays.asList(new XMLStartElement("root")),
      XMLHEADER.length() + 6, chunk1.length() - 7);
  
  /**
   * Test a simple merge
   * @param context the test context
   */
  @Test
  public void simple(TestContext context) {
    Async async = context.async();
    MergeStrategy strategy = new AllSameStrategy();
    BufferWriteStream bws = new BufferWriteStream();
    strategy.init(cm, context.asyncAssertSuccess(v1 -> {
      strategy.init(cm, context.asyncAssertSuccess(v2 -> {
        strategy.merge(new DelegateChunkReadStream(chunk1), cm, bws, context.asyncAssertSuccess(v3 -> {
          strategy.merge(new DelegateChunkReadStream(chunk2), cm, bws, context.asyncAssertSuccess(v4 -> {
            strategy.finishMerge(bws);
            context.assertEquals(XMLHEADER + "<root><test chunk=\"1\"></test><test chunk=\"2\"></test></root>",
                bws.getBuffer().toString("utf-8"));
            async.complete();
          }));
        }));
      }));
    }));
  }
  
  /**
   * Test if chunks that have not been passed to the initalize method can be merged
   * @param context the test context
   */
  @Test
  public void mergeUninitialized(TestContext context) {
    Async async = context.async();
    MergeStrategy strategy = new AllSameStrategy();
    BufferWriteStream bws = new BufferWriteStream();
    strategy.init(cm, context.asyncAssertSuccess(v1 -> {
      // skip second init
      strategy.merge(new DelegateChunkReadStream(chunk1), cm, bws, context.asyncAssertSuccess(v3 -> {
        strategy.merge(new DelegateChunkReadStream(chunk2), cm, bws, context.asyncAssertSuccess(v4 -> {
          strategy.finishMerge(bws);
          context.assertEquals(XMLHEADER + "<root><test chunk=\"1\"></test><test chunk=\"2\"></test></root>",
              bws.getBuffer().toString("utf-8"));
          async.complete();
        }));
      }));
    }));
  }
  
  /**
   * Test if canMerge works correctly
   * @param context the test context
   */
  @Test
  public void canMerge(TestContext context) {
    ChunkMeta cm2 = new ChunkMeta(Arrays.asList(new XMLStartElement("other")), 10, 20);
    ChunkMeta cm3 = new ChunkMeta(Arrays.asList(new XMLStartElement("pre", "root")), 10, 20);
    ChunkMeta cm4 = new ChunkMeta(Arrays.asList(new XMLStartElement(null, "root",
        new String[] { "" }, new String[] { "uri" })), 10, 20);
    
    Async async = context.async();
    MergeStrategy strategy = new AllSameStrategy();
    strategy.canMerge(cm, context.asyncAssertSuccess(canMerge1 -> {
      context.assertTrue(canMerge1);
      strategy.init(cm, context.asyncAssertSuccess(v1 -> {
        strategy.canMerge(cm, context.asyncAssertSuccess(canMerge2 -> {
          context.assertTrue(canMerge2);
          strategy.canMerge(cm2, context.asyncAssertSuccess(canMerge3 -> {
            context.assertFalse(canMerge3);
            strategy.canMerge(cm3, context.asyncAssertSuccess(canMerge4 -> {
              context.assertFalse(canMerge4);
              strategy.canMerge(cm4, context.asyncAssertSuccess(canMerge5 -> {
                context.assertFalse(canMerge5);
                async.complete();
              }));
            }));
          }));
        }));
      }));
    }));
  }
  
  /**
   * Test if the merge method fails if it is called with an unexpected chunk
   * @param context the test context
   */
  @Test
  public void mergeFail(TestContext context) {
    ChunkMeta cm2 = new ChunkMeta(Arrays.asList(new XMLStartElement("other")), 10, 20);
    
    Async async = context.async();
    MergeStrategy strategy = new AllSameStrategy();
    BufferWriteStream bws = new BufferWriteStream();
    strategy.init(cm, context.asyncAssertSuccess(v1 -> {
      strategy.merge(new DelegateChunkReadStream(chunk2), cm2, bws, context.asyncAssertFailure(err -> {
        async.complete();
      }));
    }));
  }
}

File: src/test/kotlin/io/georocket/output/xml/AllSameStrategyTest.kt
package io.georocket.output.xml

import io.georocket.assertThatThrownBy
import io.georocket.coVerify
import io.georocket.storage.XMLChunkMeta
import io.georocket.util.XMLStartElement
import io.georocket.util.io.BufferWriteStream
import io.vertx.core.Vertx
import io.vertx.core.buffer.Buffer
import io.vertx.junit5.VertxExtension
import io.vertx.junit5.VertxTestContext
import io.vertx.kotlin.coroutines.dispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

/**
 * Test [AllSameStrategy]
 * @author Michel Kraemer
 */
@ExtendWith(VertxExtension::class)
class AllSameStrategyTest {
  companion object {
    private const val XMLHEADER = """<?xml version="1.0" encoding="UTF-8" standalone="yes"?>${"\n"}"""
    private val chunk1 = Buffer.buffer("""<test chunk="1"></test>""")
    private val chunk2 = Buffer.buffer("""<test chunk="2"></test>""")
    private val cm = XMLChunkMeta(listOf(XMLStartElement("root")))
  }

  /**
   * Test a simple merge
   */
  @Test
  fun simple(vertx: Vertx, ctx: VertxTestContext) {
    val strategy = AllSameStrategy()
    val bws = BufferWriteStream()

    CoroutineScope(vertx.dispatcher()).launch {
      ctx.coVerify {
        strategy.init(cm)
        strategy.init(cm)
        strategy.merge(chunk1, cm, bws)
        strategy.merge(chunk2, cm, bws)
        strategy.finish(bws)

        assertThat(bws.buffer.toString("utf-8")).isEqualTo(
            """$XMLHEADER<root><test chunk="1"></test><test chunk="2"></test></root>""")
      }
      ctx.completeNow()
    }
  }

  /**
   * Test if chunks that have not been passed to the initialize method can be merged
   */
  @Test
  fun mergeUninitialized(vertx: Vertx, ctx: VertxTestContext) {
    val strategy: MergeStrategy = AllSameStrategy()
    val bws = BufferWriteStream()

    CoroutineScope(vertx.dispatcher()).launch {
      ctx.coVerify {
        strategy.init(cm) // skip second init
        strategy.merge(chunk1, cm, bws)
        strategy.merge(chunk2, cm, bws)
        strategy.finish(bws)
        assertThat(bws.buffer.toString("utf-8")).isEqualTo(
            """$XMLHEADER<root><test chunk="1"></test><test chunk="2"></test></root>""")
      }
      ctx.completeNow()
    }
  }

  /**
   * Test if canMerge works correctly
   */
  @Test
  fun canMerge(vertx: Vertx, ctx: VertxTestContext) {
    val cm2 = XMLChunkMeta(listOf(XMLStartElement("other")))
    val cm3 = XMLChunkMeta(listOf(XMLStartElement("pre", "root")))
    val cm4 = XMLChunkMeta(listOf(XMLStartElement(null, "root", arrayOf(""), arrayOf("uri"))))
    val strategy = AllSameStrategy()

    CoroutineScope(vertx.dispatcher()).launch {
      ctx.coVerify {
        assertThat(strategy.canMerge(cm)).isTrue
        strategy.init(cm)
        assertThat(strategy.canMerge(cm)).isTrue
        assertThat(strategy.canMerge(cm2)).isFalse
        assertThat(strategy.canMerge(cm3)).isFalse
        assertThat(strategy.canMerge(cm4)).isFalse
      }
      ctx.completeNow()
    }
  }

  /**
   * Test if the merge method fails if it is called with an unexpected chunk
   */
  @Test
  fun mergeFail(vertx: Vertx, ctx: VertxTestContext) {
    val cm2 = XMLChunkMeta(listOf(XMLStartElement("other")))
    val strategy = AllSameStrategy()
    val bws = BufferWriteStream()

    CoroutineScope(vertx.dispatcher()).launch {
      ctx.coVerify {
        strategy.init(cm)
        assertThatThrownBy {
          strategy.merge(chunk2, cm2, bws)
        }.isInstanceOf(IllegalStateException::class.java)
      }
      ctx.completeNow()
    }
  }
}
